packages/remix-server-runtime/server.ts

```diff
@@ -82,10 +82,9 @@ export const createRequestHandler: CreateRequestHandlerFunction = (
       );
 
       if (build.entry.module.handleDataRequest) {
-        let match = matches!.find((match) => match.route.id == routeId)!;
         response = await build.entry.module.handleDataRequest(response, {
           context: loadContext,
-          params: match ? match.params : {},
+          params: matches?.find((m) => m.route.id == routeId)?.params || {},
```

Comment:
When `handleDataRequest` was defined and a request came in like `/not-a-route?_data=routes/not-a-route` we got a valid 404 from the static handler but then errored out here trying to get the params when no matches exist.